File: Python/libraries/datatypes-timex-expression/datatypes_timex_expression/timex_date_helpers.py
from datetime import datetime, timedelta


class TimexDateHelpers:

    @staticmethod
    def tomorrow(date: datetime):
        return date.now() + timedelta(days=1)

    @staticmethod
    def yesterday(date):
        return date.now() - timedelta(days=1)

    @staticmethod
    def date_part_equal(date_x, date_y):
        return date_x == date_y

    def is_date_in_week(self, date, start_of_week):
        d = start_of_week

        for i in range(0,7,1):
            if self.date_part_equal(date, d):
                return True
            d = date.now() + timedelta(days=1)

        return False

    def is_this_week(self, date, reference_date: datetime):
        start_of_week = reference_date
        while start_of_week.weekday() > 0:
            start_of_week.now() - timedelta(days=1)
        return self.is_date_in_week(date, start_of_week)

    def is_next_week(self, date, reference_date: datetime):
        next_week_date = date.now() + timedelta(days=7)
        return self.is_this_week(date, next_week_date)

    def is_last_week(self, date, reference_date):
        next_week_date = date.now() - timedelta(days=7)
        return self.is_this_week(date, next_week_date)

    @staticmethod
    def week_of_year(date: datetime):
        ds = datetime(date.year, 1, 1)
        de = datetime(date.year, date.month, date.day)
        weeks = 1

        while ds < de:
            day_of_week = ds.weekday()

            is_day_of_week = 7 if day_of_week == 0 else int(day_of_week)

            if is_day_of_week == 7:
                weeks = weeks + 1

            ds = ds.now() + timedelta(days=1)

        return weeks

    @staticmethod
    def fixed_format_number(n, size):
        return str(n).rjust(size, '0')

    @staticmethod
    def date_of_last_day(day: datetime, reference_date: datetime):
        result = reference_date
        result = result.now() - timedelta(days=1)
        while result.weekday() != day:
            result = result.now() - timedelta(days=1)

        return result

    @staticmethod
    def date_of_next_day(day: datetime, reference_date: datetime):
        result = reference_date
        result = result.now() + timedelta(days=1)
        while result.weekday() != day:
            result = result.now() + timedelta(days=1)

        return result

    def dates_matching_day(self, day: datetime, start: datetime, end: datetime):
        result = []
        d = start
        while not self.date_part_equal(d, end):
            if d.weekday() == day:
                result.append(d)
            d = d.now() + timedelta(days=1)

        return result

File: Java/libraries/recognizers-text-datatypes-timex-expression/src/main/java/com/microsoft/recognizers/datatypes/timex/expression/TimexDateHelpers.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.recognizers.datatypes.timex.expression;

import java.time.DayOfWeek;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

public class TimexDateHelpers {
    public static Calendar tomorrow(Calendar date) {
        date.add(Calendar.DATE, 1);
        return date;
    }

    public static Calendar yesterday(Calendar date) {
        date.add(Calendar.DATE, -1);
        return date;
    }

    public static Boolean datePartEquals(Calendar dateX, Calendar dateY) {
        return (dateX.get(Calendar.YEAR) == dateY.get(Calendar.YEAR)) &&
            (dateX.get(Calendar.MONTH) == dateY.get(Calendar.MONTH)) &&
            (dateX.get(Calendar.DATE) == dateY.get(Calendar.DATE));
    }

    public static boolean isDateInWeek(Calendar date, Calendar startOfWeek) {
        Calendar d = startOfWeek;
        for (int i = 0; i < 7; i++) {
            if (TimexDateHelpers.datePartEquals(date, d)) {
                return true;
            }

            d.add(Calendar.DATE, 1);
        }

        return false;
    }

    public static Boolean isThisWeek(Calendar date, Calendar referenceDate) {
        // Note ISO 8601 week starts on a Monday
        Calendar startOfWeek = referenceDate;
        while (startOfWeek.get(Calendar.DAY_OF_WEEK) > DayOfWeek.MONDAY.getValue()) {
            startOfWeek.add(Calendar.DATE, -1);
        }

        return TimexDateHelpers.isDateInWeek(date, startOfWeek);
    }

    public static Boolean isNextWeek(Calendar date, Calendar referenceDate) {
        Calendar nextWeekDate = referenceDate;
        nextWeekDate.add(Calendar.DATE, 7);
        return TimexDateHelpers.isThisWeek(date, nextWeekDate);
    }

    public static Boolean isLastWeek(Calendar date, Calendar referenceDate) {
        Calendar nextWeekDate = referenceDate;
        nextWeekDate.add(Calendar.DATE, -7);
        return TimexDateHelpers.isThisWeek(date, nextWeekDate);
    }

    public static Integer weekOfYear(Calendar date) {
        Calendar ds = Calendar.getInstance();
        ds.set(date.get(Calendar.YEAR), 1, 1);
        Calendar de = Calendar.getInstance();
        ds.set(date.get(Calendar.YEAR), date.get(Calendar.MONTH), date.get(Calendar.DATE));
        Integer weeks = 1;

        while (ds.compareTo(de) < 0) {
            Integer dayOfWeek = ds.get(Calendar.DAY_OF_WEEK);

            Integer isoDayOfWeek = (dayOfWeek == 0) ? 7 : dayOfWeek;
            if (isoDayOfWeek == 7) {
                weeks++;
            }

            ds.add(Calendar.DATE, 1);
        }

        return weeks;
    }

    public static String fixedFormatNumber(Integer n, Integer size) {
        return String.format("%1$" + size + "s", n.toString()).replace(' ', '0');
    }

    public static Calendar dateOfLastDay(DayOfWeek day, Calendar referenceDate) {
        Calendar result = referenceDate;
        result.add(Calendar.DATE, -1);

        while (DayOfWeek.of(result.get(Calendar.DAY_OF_WEEK)) != day) {
            result.add(Calendar.DATE, -1);
        }

        return result;
    }

    public static Calendar dateOfNextDay(DayOfWeek day, Calendar referenceDate) {
        Calendar result = referenceDate;
        result.add(Calendar.DATE, 1);

        while (DayOfWeek.of(result.get(Calendar.DAY_OF_WEEK)) != day) {
            result.add(Calendar.DATE, 1);
        }

        return result;
    }

    public static List<Calendar> datesMatchingDay(DayOfWeek day, Calendar start, Calendar end) {
        List<Calendar> result = new ArrayList<Calendar>();
        Calendar d = start;

        while (!TimexDateHelpers.datePartEquals(d, end)) {
            if (DayOfWeek.of(d.get(Calendar.DAY_OF_WEEK)) == day) {
                result.add(d);
            }

            d.add(Calendar.DATE, 1);
        }

        return result;
    }
}
